From 6ea0f725b23324a5ac25b51801e63b0da6b55bec Mon Sep 17 00:00:00 2001
From: Steve Best <sbest@redhat.com>
Date: Mon, 2 Jun 2014 15:22:58 -0400
Subject: [thermal] intel_powerclamp: Fix cstate counter detection

Message-id: <20140602152258.16229.21479.sendpatchset@ibm-p7r2-01.lab.bos.redhat.com>
Patchwork-id: 81563
O-Subject: [PATCH RHEL7.1 BZ1103376 3/4] intel_powerclamp: Fix cstate counter detection.
Bugzilla: 1103376
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Mark Langsdorf <mlangsdo@redhat.com>

RHBZ#:
------
https://bugzilla.redhat.com/show_bug.cgi?id=1103376

Description:
------------
Having all zero cstate count doesn't necesserily mean the cstate
counter is no functional.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>
Signed-off-by: Zhang Rui <rui.zhang@intel.com>

RHEL Version Found:
-------------------
7.0

Upstream Status:
----------------
http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=7734e3ac89b27caf87d43d264bbbc26240413227

----------------
Steve Best

Proposed Patch:
---------------

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/thermal/intel_powerclamp.c b/drivers/thermal/intel_powerclamp.c
index 8f181b3..67b608e 100644
--- a/drivers/thermal/intel_powerclamp.c
+++ b/drivers/thermal/intel_powerclamp.c
@@ -206,6 +206,15 @@ static void find_target_mwait(void)
 
 }
 
+static bool has_pkg_state_counter(void)
+{
+ u64 tmp;
+ return !rdmsrl_safe(MSR_PKG_C2_RESIDENCY, &tmp) ||
+        !rdmsrl_safe(MSR_PKG_C3_RESIDENCY, &tmp) ||
+        !rdmsrl_safe(MSR_PKG_C6_RESIDENCY, &tmp) ||
+        !rdmsrl_safe(MSR_PKG_C7_RESIDENCY, &tmp);
+}
+
 static u64 pkg_state_counter(void)
 {
  u64 val;
@@ -500,7 +509,7 @@ static int start_power_clamp(void)
  struct task_struct *thread;
 
  /* check if pkg cstate counter is completely 0, abort in this case */
- if (!pkg_state_counter()) {
+ if (!has_pkg_state_counter()) {
   pr_err("pkg cstate counter not functional, abort\n");
   return -EINVAL;
  }
-- 
1.7.1